ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java

```diff
@@ -216,12 +245,30 @@ protected void loadTextEmbeddingModel(File modelZipFile, String modelId, String
         }
     }
 
+    private void closePredictors(Predictor[] predictors) {
+        log.debug("will close {} predictor for model {}", predictors.length, modelId);
+        for (Predictor<Input, Output> predictor : predictors) {
+            predictor.close();
+        }
+    }
+
+    private void closeModels(ZooModel[] models) {
+        log.debug("will close {} zoo model for model {}", models.length, modelId);
+        for (ZooModel model : models) {
+            model.close();
+        }
+    }
+
     protected ModelTensorOutput predictTextEmbedding(String modelId, MLInputDataset inputDataSet) {
-        log.debug("start to predict text embedding model {}", modelId);
         try {
             return AccessController.doPrivileged((PrivilegedExceptionAction<ModelTensorOutput>) () -> {
                 Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
-                if (predictor == null) {
+                int currentDevice = nextDevice.getAndIncrement();
+                if (currentDevice > devices.length - 1) {
+                    currentDevice = currentDevice % devices.length;
+                    nextDevice.set(currentDevice + 1);
+                }
```

Comment:
I think this whole section are critical needed to thread safe. For example, device length is 2, we could have two threads. Thread A has current 2, thread B has current 3. Then thread B run before thread A, then nextDevice will be set wrong number. If we are concerned the performance, how about just 
int currentDevice = nextDevice.getAndIncrement() % devices.length;

Reply:
Good point, sounds good, will change this part.